packages/firestore/src/local/web_storage.ts

```diff
@@ -0,0 +1,146 @@
+import { AsyncQueue } from '../util/async_queue';
+import { debug } from '../util/log';
+
+/**
+ * Refresh the contents of WebStorage every four seconds.
+ */
+const WEB_STORAGE_REFRESH_INTERVAL_MS: number = 4000;
+
+// Prefix keys used in WebStorage.
+const VISIBILITY_PREFIX = 'visibility';
+
+const LOG_TAG = 'WebStorage';
+
+/**
+ * WebStorage of the Firestore client. Firestore uses WebStorage for cross-tab
+ * notifications and to persist the metadata state of each tab. WebStorage is
+ * used to perform leader election and to inform other tabs about changes in the
+ * IndexedDB-backed persistence layer.
+ *
+ * WebStorage is only used in Persistence-enabled Firestore instances. If you
+ * are not using Persistence, consider using the `NoOpWebStorage` class.
+ */
+export interface WebStorage {
+  setVisibility(visibilityState: VisibilityState): void;
+  start(): void;
+  shutdown(): void;
+}
+
+/**
+ * `PersistedWebStorage` uses Local Storage as the backing store for the
+ * WebStorage class.
+ *
+ * Once started, PersistedWebStorage will rewrite its contents to Local Storage
+ * every four seconds. Other clients may disregard its state after five seconds
+ * of inactivity.
+ */
+export class PersistedWebStorage implements WebStorage {
+  private localStorage: Storage;
+  private visibilityState: VisibilityState = VisibilityState.Unknown;
+  private started = false;
+
+  constructor(
+    private persistenceKey: string,
+    private instanceId: string,
+    private asyncQueue: AsyncQueue
+  ) {}
+
+  /** Returns true if LocalStorage is available in the current environment. */
+  static isAvailable(): boolean {
+    return typeof window !== 'undefined' && window.localStorage != null;
+  }
+
+  start(): void {
+    if (!PersistedWebStorage.isAvailable()) {
+      throw new FirestoreError(
+        Code.UNIMPLEMENTED,
+        'LocalStorage is not available on this platform.'
+      );
+    }
+
+    assert(!this.started, 'PersistedWebStorage already started');
+
+    this.localStorage = window.localStorage;
+    this.started = true;
+    this.persistState();
+    this.scheduleRefresh();
+  }
+
+  shutdown(): void {
+    assert(
+      this.started,
+      'PersistedWebStorage.shutdown() called when not started'
+    );
+    this.started = false;
+  }
+
+  setVisibility(visibilityState: VisibilityState): void {
+    this.visibilityState = visibilityState;
+    this.persistState();
+  }
+
+  private scheduleRefresh(): void {
+    this.asyncQueue.schedulePeriodically(() => {
+      if (this.started) {
+        this.persistState();
+      }
+      return Promise.resolve();
+    }, WEB_STORAGE_REFRESH_INTERVAL_MS);
+  }
+
+  /** Persists the entire known state. */
+  private persistState(): void {
+    assert(this.started, 'PersistedWebStorage not started');
+    debug(LOG_TAG, 'Persisting state in LocalStorage');
+    this.localStorage[
+      this.buildKey(VISIBILITY_PREFIX, this.persistenceKey, this.instanceId)
```

Comment:
This key seems like something you could build once in the constructor and not even retain the arguments used to construct it.

Reply:
Sounds good.